### DISCUSSION TITLE
Mixed colour/white mode Tuya lamp

### DISCUSSION THREAD

**Discussion by akantifirst** [jasonacox/tinytuya#194] [Q&A] [answered]
I am able to adjust the parameters of a Tuya "Type B" ceiling lamp from Tuya IoT platform by sending the following JSON:

`[
        {
            "code":"control_data",
            "value":{"change_mode":"gradient",
                "bright": 999,
                "temperature":0,
                "h":75,
                "s":999,
                "v":999
            }
        }
    ]`
    
   However trying to send the same command as payload using Tinytuya switches the lamp either to white or to colour mode:

`command = {'20': True, '21': 'scene', '22': 999, '23': 0, '24': '01c901ff0383', '26': 0}
payload = d.generate_payload(tinytuya.CONTROL, command)
d.send(payload)`

It seems that the lamp works in two modes and is not possible to send one command that adjusts RGB as well as white light at the same time. 
Had anyone faced such as a problem?

**Comment by akantifirst** [comment]
I have managed to solve this problem in the following way.
On the **Tuya IoT Platform** in the _Device Debugging_ section we can send instructions using JSON Editor:

![image](https://user-images.githubusercontent.com/82590770/195050592-899a172c-6618-4406-b06a-5c31efc29076.png)

After successfully sending the instruction we go to the _Device Logs_ section and see there a new  _Adjustment_ record. 
In the _Event Details_ column there is a string that we are going to copy. 

![image](https://user-images.githubusercontent.com/82590770/195050913-56099879-1fa5-4886-be7f-7267d353a2a3.png)

Now, from python script we can adjust the lamp brightness, temperature, and RGB parameters in one line of code:

`d.set_value('28', '10000000003e803e803e8')`

**'28'** is Debugging mode DPS ID

Despite this code is working fine, in Pycharm I get warning that `int` and not `str` is expected as argument.

![image](https://user-images.githubusercontent.com/82590770/195051177-301eec0a-d424-40fd-a3eb-08aa8530e9c0.png)

However, it is still not clear how to generate such a string with parameters like brightness, temparature, RGB color etc. 
Also, it is not clear how to encode transitions in such strings. For example if you want your lamps to light up slowly with transition duration 2-3 seconds.

**Reply by uzlonewolf** [reply]
That "Mix Light" DPS may be needed for the transition, I'm not sure.  The base64 value above when decoded and converted into hex is `0007003703bf036b02a903d5` - assuming they're all 16-bit numbers it would be `7 55 959 875 681 981` when converted to decimal.  Your best bet is to send a transition command from the app and see what those device logs show.

**Reply by akantifirst** [reply]
Thanks for advice that this string is base64-encoded! Now I am able to program my own scene / transitions direct in code.

**Comment by akantifirst** [comment, marked as answer]
For anybody interested, this is the way to send **mixed light scene** commands (_white_ LED and _colour_ LED are controlled at the same time). I have made the illustration below to show how differerent parameters are encoded in control string: 
![image_2022-10-11_17-13-19](https://user-images.githubusercontent.com/82590770/197159104-7f6296f0-febf-4b60-81ef-b59ddbe246f4.png)

Tested with _Torkase Alexa Smart Ceiling Light, 35 W 2400 LM WiFi Dimmable Ceiling Light Controllable via App_

![image](https://user-images.githubusercontent.com/82590770/197159976-2acaca9d-4fd2-4cba-9c57-f4f90f321da0.png)

However, the `'28'` "debugger" DPS is useless once the `switch_led` parameter is set to `False`, for example, after the `countdown` timer fires. The lamp becomes unresponsive.

After some experimenting with _Smart Industry App_ and looking in the debugging logs on _Tuya IoT Platform_, I have found 2 extra, non-documented DPS `36` for _Mix Light Scene_ and `51` for _Mix Light_ :

![image](https://user-images.githubusercontent.com/82590770/197162863-018bb066-780e-4893-b9be-c765f9e0f416.png)

These are base64-encoded values. I don't know how to decode their meaning yet, but I suppose that `36` is for light parameters and `51` is for transition parameters. These DPS's seems to be more "powerful" than the `28` _debugger_ DPS. 
This is how it is possible to send mixed light scene command to the lamp **even if the `switch_led` parameter is set to `False`**:

```
d.set_value(36, mix_light_scene)
d.set_value(51, mix_light)
```

I am finally happy to have full local control of my 7x Tuya Ceiling Lights, even without the need to resolder chips and flash Tasmota. Big thank to the author of this awesome library and contributors!
